### PR TITLE
Add documentation on goto-analyzer to the manual

### PR DESCRIPTION
**Pull request by martin-cs**
I have promised this to (in order of oldest promise) @tautschnig @remi-delmas-3000 @jimgrundy and @vmihalko .  If any of you have a chance to look at this and say whether it is what you wanted / needed / hoped for then I would be very grateful.

### CODE REVIEW

File: doc/cprover-manual/goto-analyzer.md [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+with CBMC, so [the installation instructions](../installation/) are
+the same.
+
+The key difference is that CBMC under-approximates the
+behaviour of the program (execution traces that are too long or
+require too many loop unwindings are not considered) while
+`goto-analyzer` over-approximates the behaviour of the program.  CBMC
+can determine if a property is A. true for a bounded number of
+iterations or B. false and giving an error trace.  In contrast
+`goto-analyzer` can determin if a property is A. true for all
```
Comment (peterschrammel) [inline doc/cprover-manual/goto-analyzer.md]:
```suggestion
`goto-analyzer` can determine if a property is A. true for all
```

File: doc/cprover-manual/goto-analyzer.md [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+
+
+
+### Task {#task}
+
+`goto-analyzer` first runs the abstract interpreter until it reaches a
+fix-point, then it will perform the task the user has chosen.
+
+`--show`
+: Displays a domain for every instruction in the GOTO binary.  The
```
Comment (peterschrammel) [inline doc/cprover-manual/goto-analyzer.md:68]: 🦁 domain or abstract value?

File: doc/cprover-manual/goto-analyzer.md [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+format and information will depend on the [domain](#domain) that has
+been selected.  If there are multiple domains corresponding to the
```
Comment (peterschrammel) [inline doc/cprover-manual/goto-analyzer.md:70]: Multiple domains in the sense of domain product, for instance, or multiple abstract values?

File: doc/cprover-manual/goto-analyzer.md [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+same location these will be merged before they are displayed.
+
+`--show-on-source`
+: The source code of the program is displayed line-by-line with the
+abstract domains corresponding to each location displayed between
```
Comment (peterschrammel) [inline doc/cprover-manual/goto-analyzer.md:76]: 🦁

File: doc/cprover-manual/goto-analyzer.md [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+them.  As the analysis is done at the level of instructions in the
+GOTO binary, some domains may not be displayed.  Also if parts of the
+GOTO binary have been generated or manipulated by other tools, these
+may not be displayed as there is no corresponding source.
+`--show-on-source` is the more user-friendly output but `--show` gives
```
Comment (peterschrammel) [inline doc/cprover-manual/goto-analyzer.md]:
```suggestion
`--show-on-source` is the more user-friendly output, but `--show` gives
```

File: doc/cprover-manual/goto-analyzer.md [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+deactivated by passing `--no-simplify-slicing`.  In the ideal world
+simplify would be idempotent (i.e. running it a second time would not
+simply anything more than the first).  However there are edge cases
+which are difficult or prohibitatively expensive to handle in the
+domain which can result in a second (or more) runs giving
+simplification.  Submitting bug reports for these is helpful but they
+may not be viable to fix.
+
+`--unreachable-instructions`
+: Lists which instructions have a domain which is bottom
```
Comment (peterschrammel) [inline doc/cprover-manual/goto-analyzer.md:109]: 🦁

File: doc/cprover-manual/goto-analyzer.md [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+: This extends `--legacy-ait` by allowing the History and Storage to
+be configured.  As the name implies, function calls are handled by
+recursion within the interpreter.  This is a good all-round choice and
+will likely become the default at some point in the future.
+
+`--three-way-merge`
+: This extends `--recursive-interprocedural` by performing a
+"modification aware" merge after function calls.  At the time of
+writing only `--vsd` supports the necessary differential reasoning.
+If you are using `--vsd` this is recommended as it is more accurate at
```
Comment (peterschrammel) [inline doc/cprover-manual/goto-analyzer.md]:
```suggestion
If you are using `--vsd` this is recommended as it is more accurate 
```

File: doc/cprover-manual/goto-analyzer.md [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+only `--dependence-graph` is compatible with it.
+
+`--recursive-interprocedural`
+: This extends `--legacy-ait` by allowing the History and Storage to
+be configured.  As the name implies, function calls are handled by
+recursion within the interpreter.  This is a good all-round choice and
+will likely become the default at some point in the future.
+
+`--three-way-merge`
+: This extends `--recursive-interprocedural` by performing a
```
Comment (peterschrammel) [inline doc/cprover-manual/goto-analyzer.md:182]: ❓ So, everything is derived from "legacy", i.e. everything is legacy then? ;)

File: doc/cprover-manual/goto-analyzer.md [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+: A domain that stores an interval for each integer and float
+variable.  At the time of writing not all operations are supported so
+the results can be quite over-approximate at points.  It also has
+limitations in the handling of pointers so can give unsound results.
+`--vsd --vsd-values intervals` is probably a better choice for this
+kind of analysis.
+
+`--not-null`
+: This domain is intended to find which pointers are not null.  Its
+implementation is very limited an it is not recommended.
```
Comment (peterschrammel) [inline doc/cprover-manual/goto-analyzer.md]:
```suggestion
implementation is very limited and it is not recommended.
```

File: doc/cprover-manual/goto-analyzer.md [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+
+`--dependence-graph`
+: Tracks data flow and information flow dependencies between
+instructions and produces a graph.  This includes doing points-to
+analysis and tracking reaching definitions (i.e. use-def chains).
+This is one of the most extensive, correct and feature complete domains.
+
+`--vsd`
+: This is the Variable Sensitivity Domain (VSD).  It is a non-relational
+domain that stores an abstract object for each live variable.  Which
```
Comment (peterschrammel) [inline doc/cprover-manual/goto-analyzer.md:222]: abstract object 👍

File: doc/cprover-manual/goto-analyzer.md [hunk 11]
```diff
@@ -0,0 +1,432 @@
...
+
+`--vsd-unions`
+: At the time of writing there is only one option for unions which is
+`top-bottom`, discarding writes and returning top for all reads from a
+variable of union type.
+
+`--vsd-data-dependencies`
+: Wraps each abstract object with a set of locations whether the
+variable was last modified.  The set is reset when the variable is
+written and takes the union of the two side's sets on merge.  This was
```
Comment (peterschrammel) [inline doc/cprover-manual/goto-analyzer.md]:
```suggestion
written and takes the union of the two sides' sets on merge.  This was
```

File: doc/cprover-manual/goto-analyzer.md [hunk 12]
```diff
@@ -0,0 +1,432 @@
...
+expensive in terms of both time and memory but can give much more
+accurate results.  Recursive functions create a challenge as the call
+stack will be different each time.  To prevent non-termination, the
+parameter `n` limits how many times a loop of recursive functions can
+be called.  When `n` is reached all later ones will be merged.
+Setting this to `0` will disable the limit.
+
+`--loop-unwind n`
+: This tracks the backwards jumps that are taken in the current
+function.  Traces that reach the same location are merged of their
```
Comment (peterschrammel) [inline doc/cprover-manual/goto-analyzer.md]:
```suggestion
function.  Traces that reach the same location are merged if their
```

File: doc/cprover-manual/goto-analyzer.md [hunk 13]
```diff
@@ -0,0 +1,432 @@
...
+histories match. This gives a similar effect to unrolling the loops
+`n` times and then using `--ahistorical`.  In the case of nested
+loops, the behaviour can be a little different to unrolling as the
+limit is the number of times a location is reached, so a loop with `x`
+iterations containing a loop with `y` iterations will require `n = x*y`.
+The time and memory taken by this option will rise (at worst) linearly
+in terms of `n`.  If `n` is `0` then there is no limit.  Be warned
+that if there are loops that can execute an unbounded number of
+iterations or if the domain is not sufficiently precise to identify
+the termination conditions then the analysis then this will not terminate.
```
Comment (peterschrammel) [inline doc/cprover-manual/goto-analyzer.md]:
```suggestion
the termination conditions then the analysis will not terminate.
```

File: doc/cprover-manual/goto-analyzer.md [hunk 14]
```diff
@@ -0,0 +1,432 @@
...
+each function.
+
+
+
+
+### Storage {#storage}
+
+The histories described above are used to keep track of where in the
+computation needs to be explored.  The most precise option is to keep
+one domain for every history but in some cases, to save memory and
```
Comment (peterschrammel) [inline doc/cprover-manual/goto-analyzer.md:405]: 🦁

**Review by peterschrammel — APPROVED** [review]

**Comment by martin-cs** [comment]
@peterschrammel thank you for the review and feedback; very helpful.  What do you mean by ":lion:"?

**Comment by chrisr-diffblue** [comment]
> @peterschrammel thank you for the review and feedback; very helpful. What do you mean by "🦁"?

I think its a reference back to one of his previous comments "🦁 domain or abstract value?" - i.e. that comment applies everywhere else he's used the same emoij.

**Comment by martin-cs** [comment]
Aaaaahhhhhhh; thank you @chrisr-diffblue

**Comment by chrisr-diffblue** [comment]
Either that, or its a warning you are pulling a lions tail 😁

**Comment by jimgrundy** [comment]
@martin-cs Thank you!

**Review by tautschnig — APPROVED** [review]
Thank you!!

File: doc/cprover-manual/goto-analyzer.md (hunk 2 above)
Comment (martin-cs) [reply doc/cprover-manual/goto-analyzer.md:68]: It displays the whole `ai_domaint` object.  I am referring to this as the domain. In cases (in VSD) where I am talking about single `abstract_objectt`'s I am calling them abstract values.  Do you feel this is the correct terminology?

File: doc/cprover-manual/goto-analyzer.md (hunk 3 above)
Comment (martin-cs) [reply doc/cprover-manual/goto-analyzer.md:70]: Multiple domains from multiple histories, I have changed the wording to make this clearer.

File: doc/cprover-manual/goto-analyzer.md (hunk 4 above)
Comment (martin-cs) [reply doc/cprover-manual/goto-analyzer.md:76]: As with the original :lion:

File: doc/cprover-manual/goto-analyzer.md (hunk 14 above)
Comment (martin-cs) [reply doc/cprover-manual/goto-analyzer.md:405]: As the original :lion: above.

**Comment by martin-cs** [comment]
@peterschrammel only the :lion: issues remain then I can merge.  Let me know!

**Review by jimgrundy — APPROVED** [review]
This is great. For those options that are unsound, can we get some warning in the output too incase someone uses them without reading the manual (perish the thought).

**Review by peterschrammel — APPROVED** [review]
Fine with me

**Comment by martin-cs** [comment]
@tautschnig @peterschrammel @chris-ryder The only failing tests are from the code coverage which has dropped because documentation seems to be included in the "percentage covered" metric.  Do you mind if I just merge it?

**Comment by tautschnig** [comment]
> @tautschnig @peterschrammel @chris-ryder The only failing tests are from the code coverage which has dropped because documentation seems to be included in the "percentage covered" metric. Do you mind if I just merge it?

No concerns from my side!

**Comment by martin-cs** [comment]
Thanks; I'll merge and if anyone knows how to remove documentation from the coverage count (or how we should test the documentation?) then I will have a go at fixing that in another PR.

**Pull request merged**